### PR TITLE
Add per-backend optimization_level config (GH-1322)

### LINKED ISSUE #1322: [REQ] Separate CUDA and CPU compilation configs
### Description

Some of Warp's `config` options currently control both the compilation of CUDA kernels and CPU kernels. We should consider splitting them for more control when they don't quite line up and users may want different behavior for each type of kernel.

Perhaps it could even distinguish between different CUDA devices. For example `wp.config.optimization_level["cuda:1"] = 2` when it provides an advantage for that specific GPU.

### Context

#1310 illustrates that while initially it appeared desirable compile CPU kernels at optimization level 3 (just like CUDA kernels), it turned out level 2 is a more balanced default. While using `config.optimization_level=None` can accommodate that by using different effective defaults, it means it's currently not possible to compile CUDA kernels at level 2 while compiling CPU kernels at level 3. Perhaps it can be done by switching the config in between launches, but this would be very fragile.

Note that kernels with `modules="unique"` could also take overrides for config settings.

### PR DESCRIPTION
**Pull request by devansh0703**
## Description

Allow `wp.config.optimization_level` to accept a dict for independent CPU and CUDA optimization levels, with optional per-device CUDA overrides. Previously the setting applied the same int to both backends, making it impossible to compile CPU at O3 while CUDA at O2 (or vice versa).

Closes #1322

## Changes

- **`warp/config.py`**: `optimization_level` type broadened to `int | None | dict`. Added `_validate_optimization_level()` for assignment-time validation and `_resolve_optimization_level()` to resolve per-backend at compile time. Supports `"cuda:N"` keys for per-device overrides.
- **`warp/_src/context.py`**: `_compile()` resolves dict-valued optimization_level before passing the int to `build_cpu()` / `build_cuda()`.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://nvidia.github.io/warp/stable/user_guide/contribution_guide.html).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] [CHANGELOG.md](CHANGELOG.md) is updated for any user-facing changes under the `Unreleased` section.

## Validation summary

- All 34 existing unittests pass (`test_options`, `test_module_hashing`) from both warm and cold cache.
- Dict resolution verified at every layer: `config attribute → resolve_options → _resolve_optimization_level → build_cpu/build_cuda`.
- Integration test with `{"cpu": 0, "cuda": 3}` confirmed correct compilation and execution on both backends.
- `module_options` dict override (both module-level and `module="unique"`) verified.
- Edge cases: invalid keys/values rejected, `None` in dict, per-device fallback to `"cuda"` key, hash distinctness.

## New feature / enhancement

```python
import warp as wp

# Per-backend: CPU at O2, CUDA at O3
wp.config.optimization_level = {"cpu": 2, "cuda": 3}

# Per-device: cuda:0 overrides the general CUDA setting
wp.config.optimization_level = {"cpu": 2, "cuda": 3, "cuda:0": 1}

# Per-kernel with module="unique"
@wp.kernel(module_options={"optimization_level": {"cpu": 0, "cuda": 3}}, module="unique")
def debug_kernel(...):
    ...

# Backward-compatible: int and None still work as before
wp.config.optimization_level = 2
wp.config.optimization_level = None


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Expanded `optimization_level` to accept dictionary-based settings for per-backend configuration (e.g., `"cpu"`, `"cuda"`) with per-CUDA-device overrides via `"cuda:N"`.
* **Bug Fixes**
  * Added stricter validation for allowed dictionary keys and value types/ranges.
  * Ensures the effective optimization level is resolved consistently, prioritizing `"cuda:N"` over generic `"cuda"` for the active device.
* **Tests**
  * Added coverage for dict-based `optimization_level`, including execution and hash differentiation.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

**Comment by devansh0703** [comment]
Addressed in 10824634:

- **CodeRabbit (actionable):** Moved validation into `Module.resolve_options()`, the common compile-time sink that all paths (`wp.config`, `set_module_options()`, `@wp.kernel(module_options=...)`) route through. Added validation call and Nones handling (the None fallback sets the dict from config, validation runs after).

- **Greptile (range):** Added 0-3 range check for dict int values in `_validate_optimization_level()`.

- **Greptile (hash order):** Normalize dict to sorted keys in `resolve_options()` so hash is deterministic regardless of insertion order.

**Comment by devansh0703** [comment]
Addressed Copilot review in 69b0f9a:

- **Non-str keys:** Added explicit key type check before `.startswith()` -> clear `ValueError`
- **Type annotation:** `dict` -> `dict[str, int | None]`
- **Bool rejection:** `isinstance(value, bool)` check before int check (bool is subclass of int)
- **CHANGELOG:** Double backticks -> single backticks for markdown consistency
- **Tests added:** `test_options_opt_level_dict`, `test_options_opt_level_dict_hash`, `test_options_opt_level_dict_per_device`
All 28 tests passing.

File: warp/_src/context.py [hunk 1]
```diff
@@ -3698,9 +3721,22 @@ def _compile(
         else:
             arch_suffix = ""
 
+        # Resolve per-device optimization level early so the cache key
+        # (output filename) differs when two same-arch devices have
+        # different per-device overrides.
+        opt = options["optimization_level"]
+        if isinstance(opt, dict):
+            opt = warp.config._resolve_optimization_level(opt, is_cpu, device)
+            options["optimization_level"] = opt  # replace dict with resolved int
```
[the remark this answers is not in the document]
Comment (devansh0703) [reply warp/_src/context.py:3737]: Not a bug — `_compile()` rebinds `options` at line 3680 via `options = options | {"output_arch": output_arch}`, which creates a new dict (Python's `|` operator is equivalent to `{**a, **b}`). The subsequent `options["optimization_level"] = opt` mutation at line 3730 only affects this ephemeral merge result, never the original `self.resolved_options[active_block_dim]` entry. The shared cache dict is never mutated.
